From 4bba83fbc6ac13ef55259a1f6241ab3d30f73e42 Mon Sep 17 00:00:00 2001
From: sdev <sdev@75f.io>
Date: Thu, 16 Jun 2022 05:49:25 +0000
Subject: [PATCH 02/35] Merged PR 4135: 12285 : CCU | BPOS| zone is not
 reverting to temporary hold (auto)

BPOS algorithm correction

Related work items: #12285
---
 logic/src/main/java/a75f/io/logic/L.java      |   1 +
 .../logic/bo/building/bpos/BPOSProfile.java   | 267 ++++++++----------
 .../io/logic/jobs/ScheduleProcessJob.java     |   2 +-
 3 files changed, 117 insertions(+), 153 deletions(-)

diff --git a/logic/src/main/java/a75f/io/logic/L.java b/logic/src/main/java/a75f/io/logic/L.java
index 959b3ab71..d53f3e5a2 100644
--- a/logic/src/main/java/a75f/io/logic/L.java
+++ b/logic/src/main/java/a75f/io/logic/L.java
@@ -53,6 +53,7 @@ public class L
     public static final String TAG_CCU_MESSAGING = "CCU_MESSAGING";
     public static final String TAG_CCU_WEATHER = "CCU_WEATHER";
     public static final String TAG_CCU_MIGRATION_UTIL = "MIGRATION_UTIL";
+    public static final String TAG_BPOS = "CCU_BPOS";
 
     public static Context app()
     {
diff --git a/logic/src/main/java/a75f/io/logic/bo/building/bpos/BPOSProfile.java b/logic/src/main/java/a75f/io/logic/bo/building/bpos/BPOSProfile.java
index 13d5730f8..261ae4fd3 100644
--- a/logic/src/main/java/a75f/io/logic/bo/building/bpos/BPOSProfile.java
+++ b/logic/src/main/java/a75f/io/logic/bo/building/bpos/BPOSProfile.java
@@ -2,17 +2,13 @@ package a75f.io.logic.bo.building.bpos;
 
 import android.util.Log;
 
-import com.fasterxml.jackson.annotation.JsonIgnore;
-
-import org.joda.time.DateTime;
 import org.projecthaystack.HNum;
 import org.projecthaystack.HRef;
 
-import java.sql.Time;
-import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Objects;
 import java.util.Set;
 
 import a75f.io.api.haystack.CCUHsApi;
@@ -21,36 +17,26 @@ import a75f.io.api.haystack.HSUtil;
 import a75f.io.api.haystack.HisItem;
 import a75f.io.api.haystack.Occupied;
 import a75f.io.api.haystack.Point;
-import a75f.io.api.haystack.Schedule;
 import a75f.io.logger.CcuLog;
-import a75f.io.logic.Globals;
 import a75f.io.logic.L;
-import a75f.io.logic.bo.building.BaseProfileConfiguration;
 import a75f.io.logic.bo.building.Occupancy;
 import a75f.io.logic.bo.building.ZoneProfile;
 import a75f.io.logic.bo.building.definitions.ProfileType;
 import a75f.io.logic.bo.building.system.SystemMode;
 import a75f.io.logic.bo.building.system.dab.DabSystemController;
-import a75f.io.logic.bo.haystack.device.SmartNode;
 import a75f.io.logic.bo.util.SystemTemperatureUtil;
 import a75f.io.logic.jobs.ScheduleProcessJob;
 import a75f.io.logic.jobs.SystemScheduleUtil;
-import a75f.io.logic.tuners.BPOSTuners;
-import a75f.io.logic.tuners.StandaloneTunerUtil;
 import a75f.io.logic.tuners.TunerUtil;
 
 import static a75f.io.logic.bo.building.ZoneState.COOLING;
-import static a75f.io.logic.bo.building.ZoneState.DEADBAND;
 import static a75f.io.logic.bo.building.ZoneState.HEATING;
 import static a75f.io.logic.bo.building.ZoneState.TEMPDEAD;
 
 /*
  * created by spoorthidev on 3-August-2021
  */
-
 public class BPOSProfile extends ZoneProfile {
-
-
     BPOSEquip mBPOSEquip;
     public HashMap<Integer, BPOSEquip> mBposDeviceMap;
 
@@ -100,14 +86,14 @@ public class BPOSProfile extends ZoneProfile {
                 "auto and forced and occupied and config and equipRef == \"" + mBPOSEquip.mEquipRef + "\"") > 0;
         boolean isAutoawayenabled = CCUHsApi.getInstance().readDefaultVal("point and " +
                 "auto and forced and away and config and equipRef == \"" + mBPOSEquip.mEquipRef + "\"") > 0;
-        HashMap occupancy = CCUHsApi.getInstance().read("point and occupancy and mode and " +
+        HashMap<Object,Object> occupancy = CCUHsApi.getInstance().readEntity("point and occupancy and mode and " +
                 "equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
         double occupancyvalue =
-                CCUHsApi.getInstance().readHisValById(occupancy.get("id").toString());
-    
-    
+                CCUHsApi.getInstance().readHisValById(Objects.requireNonNull(occupancy.get("id")).toString());
+
+
         double forcedOccupiedMinutes = TunerUtil.readTunerValByQuery("forced and occupied and time",
-                                                                  mBPOSEquip.mEquipRef);
+                mBPOSEquip.mEquipRef);
         //We don't need to update forced occupancy if the tuner is 0.
         if (isAutoforceoccupiedenabled && forcedOccupiedMinutes > 0) {
             runAutoForceOccupyOperation(mBPOSEquip.mEquipRef);
@@ -120,37 +106,32 @@ public class BPOSProfile extends ZoneProfile {
             resetForceOccupy();
         }
 
-        double zoneCurTemp = mBPOSEquip.getCurrentTemp();
         double desiredTempCooling =
                 SystemTemperatureUtil.getDesiredTempCooling(mBPOSEquip.mEquipRef);
         double desiredTempHeating =
                 SystemTemperatureUtil.getDesiredTempHeating(mBPOSEquip.mEquipRef);
         double tempMidPoint = (desiredTempCooling + desiredTempHeating) / 2;
         mBPOSEquip.setDesiredTemp(tempMidPoint);
-        double zoneCoolingLoad = zoneCurTemp > tempMidPoint ? zoneCurTemp - desiredTempCooling : 0;
-        double zoneHeatingLoad = zoneCurTemp < tempMidPoint ? desiredTempHeating - zoneCurTemp : 0;
 
         String zoneId = HSUtil.getZoneIdFromEquipId(mBPOSEquip.mEquipRef);
         Occupied occ = ScheduleProcessJob.getOccupiedModeCache(zoneId);
-        boolean occupied = (occ == null ? false : occ.isOccupied());
+        boolean occupied = (occ != null && occ.isOccupied());
 
 
 
 
-        HashMap occDethashmap = CCUHsApi.getInstance().read("point and occupancy and " +
+        HashMap<Object,Object> occDethashmap = CCUHsApi.getInstance().readEntity("point and occupancy and " +
                 "detection and his and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
-        double zonedynamicpriority =
-                DabSystemController.getInstance().getEquipDynamicPriority(zoneCoolingLoad != 0 ?
-                        zoneCoolingLoad : zoneHeatingLoad, mBPOSEquip.mEquipRef);
         double occDetPoint =
-                CCUHsApi.getInstance().readHisValById(occDethashmap.get("id").toString());
+                CCUHsApi.getInstance().readHisValById(Objects.requireNonNull(occDethashmap.get(
+                        "id")).toString());
         double setTempCooling = mBPOSEquip.getDesiredTempCooling();
         double setTempHeating = mBPOSEquip.getDesiredTempHeating();
         double roomTemp = mBPOSEquip.getCurrentTemp();
         double systemDefaultTemp = 72.0;
 
 
-        Log.d("BPOSProfile", "occupied = " + occupied
+        Log.d(L.TAG_BPOS, "occupied = " + occupied
                 + "isAutoforceoccupiedenabled = " + isAutoforceoccupiedenabled
                 + "isAutoawayenabled = " + isAutoawayenabled
                 + "occupancyvalue = " + occupancyvalue
@@ -189,7 +170,7 @@ public class BPOSProfile extends ZoneProfile {
         mBPOSEquip.setStatus(state.ordinal(),
                 DabSystemController.getInstance().isEmergencyMode() && (state == HEATING ?
                         buildingLimitMinBreached()
-                        : state == COOLING ? buildingLimitMaxBreached() : false));
+                        : state == COOLING && buildingLimitMaxBreached()));
 
     }
 
@@ -206,12 +187,8 @@ public class BPOSProfile extends ZoneProfile {
                 "buildingLimitMax:" + buildingLimitMax + " tempDead:" + tempDeadLeeway);
         CcuLog.d(L.TAG_CCU_ZONE, " roomTemp : " + mBPOSEquip.getCurrentTemp() + " " +
                 "buildingLimitMin:" + buildingLimitMin + " tempDead:" + tempDeadLeeway);
-        if (mBPOSEquip.getCurrentTemp() > (buildingLimitMax + tempDeadLeeway)
-                || mBPOSEquip.getCurrentTemp() < (buildingLimitMin - tempDeadLeeway)) {
-            return true;
-        }
-
-        return false;
+        return mBPOSEquip.getCurrentTemp() > (buildingLimitMax + tempDeadLeeway)
+                || mBPOSEquip.getCurrentTemp() < (buildingLimitMin - tempDeadLeeway);
     }
 
     @Override
@@ -240,21 +217,16 @@ public class BPOSProfile extends ZoneProfile {
                 "point and  bpos and occupancy  and his and " +
                         "sensor and equipRef  == \"" + mBPOSEquip.mEquipRef + "\"") > 0;
 
-        HashMap occupancymode = CCUHsApi.getInstance().read(
-                "point and occupancy and mode and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
-
-
         Occupied occuStatus =
                 ScheduleProcessJob.getOccupiedModeCache(HSUtil.getZoneIdFromEquipId(equipRef));
 
-        Log.d("BPOSProfile", "occupied = " + occuStatus
+        Log.d(L.TAG_BPOS, "occupied = " + occuStatus
                 + "occupancyModeval = " + occupancyModeval
                 + "occupancysensor = " + occupancysensor);
 
+        assert occuStatus != null;
         if ((!occuStatus.isOccupied() || occuStatus.getVacation() != null)
-                && occupancyModeval != Occupancy.AUTOAWAY.ordinal()
-        ) {
-
+                && occupancyModeval != Occupancy.AUTOAWAY.ordinal()) {
             long temporaryHoldTime =
                     ScheduleProcessJob.getTemporaryHoldExpiry(HSUtil.getEquipInfo(mBPOSEquip.mEquipRef));
             long differenceInMinutes = findDifference(temporaryHoldTime, true);
@@ -278,26 +250,26 @@ public class BPOSProfile extends ZoneProfile {
                                     "equipRef == \"" + mBPOSEquip.mEquipRef + "\"",
                             (double) Occupancy.AUTOFORCEOCCUPIED.ordinal());
 
-                    updateDesiredtemp(desiredAvgTemp);
-                    Log.i("BPOS", "Falling in FORCE Occupy mode");
+                    updateDesiredtemp(desiredAvgTemp,true);
+                    Log.i(L.TAG_BPOS, "Falling in FORCE Occupy mode");
                 } else {
-                    Log.i("BPOS", "We are already in force occupy");
+                    Log.i(L.TAG_BPOS, "We are already in force occupy");
                     // We are already in force occupy
                     // Just update with latest
-                    updateDesiredtemp(desiredAvgTemp);
+                    updateDesiredtemp(desiredAvgTemp,false);
                 }
             } else {
-                Log.d("BPOSProfile", "Occupant not detected in unoccupied mode");
+                Log.d(L.TAG_BPOS, "Occupant not detected in unoccupied mode");
                 if (occupancyModeval == Occupancy.AUTOFORCEOCCUPIED.ordinal() && differenceInMinutes <= 0) {
                     resetForceOccupy();
                 }
             }
         } else {
             // Reset everything if there was force occupied condition
-            Log.d("BPOSProfile", "We are back to occupied");
-            Log.d("BPOSProfile", "occupancyModeval = "+occupancyModeval);
+            Log.d(L.TAG_BPOS, "We are back to occupied");
+            Log.d(L.TAG_BPOS, "occupancyModeval = "+occupancyModeval);
             if (occupancyModeval == Occupancy.AUTOFORCEOCCUPIED.ordinal()) {
-                Log.d("BPOSProfile", "Move to to occupied status");
+                Log.d(L.TAG_BPOS, "Move to to occupied status");
                 resetForceOccupy();
             }
         }
@@ -313,94 +285,83 @@ public class BPOSProfile extends ZoneProfile {
     }
 
     private void resetForceOccupy() {
-        Log.i("BPOS", "Resetting the resetForceOccupy: ");
+        Log.d(L.TAG_BPOS, "Resetting the resetForceOccupy: ");
         CCUHsApi.getInstance().writeHisValByQuery("point and  bpos and occupancy  and his and " +
                 "mode and equipRef == \"" + mBPOSEquip.mEquipRef + "\"", 0.0);
         CCUHsApi.getInstance().writeHisValByQuery("point and  bpos and occupancy  and his and " +
                         "mode and equipRef == \"" + mBPOSEquip.mEquipRef + "\"",
                 (double) Occupancy.UNOCCUPIED.ordinal());
 
-        HashMap coolDT = CCUHsApi.getInstance().read("point and desired and cooling and equipref " +
+        HashMap<Object,Object> coolDT = CCUHsApi.getInstance().readEntity("point and desired and cooling and equipref " +
                 "== \"" + mBPOSEquip.mEquipRef + "\"");
-        HashMap heatDT = CCUHsApi.getInstance().read("point and desired and heating and equipref " +
+        HashMap<Object,Object> heatDT = CCUHsApi.getInstance().readEntity("point and desired and heating and equipref " +
                 "== \"" + mBPOSEquip.mEquipRef + "\"");
 
-        SystemScheduleUtil.clearOverrides(heatDT.get("id").toString());
-        SystemScheduleUtil.clearOverrides(coolDT.get("id").toString());
+        SystemScheduleUtil.clearOverrides(Objects.requireNonNull(heatDT.get("id")).toString());
+        SystemScheduleUtil.clearOverrides(Objects.requireNonNull(coolDT.get("id")).toString());
     }
 
-    private void updateDesiredtemp(Double dt) {
-        Log.d("BPOSProfile", "  updateDesiredtemp ++");
-        HashMap equipMap =
-                CCUHsApi.getInstance().read("equip and group == \"" + mBPOSEquip.mNodeAddr + "\"");
+    private void updateDesiredtemp(Double avgDesiredTemp,boolean isFirstTime) {
+        Log.d(L.TAG_BPOS, "  updateDesiredtemp ++");
+        CCUHsApi hayStack = CCUHsApi.getInstance();
+        HashMap<Object,Object> equipMap =
+                hayStack.readEntity("equip and group == \"" + mBPOSEquip.mNodeAddr + "\"");
         Equip q = new Equip.Builder().setHashMap(equipMap).build();
 
-        double cdb = StandaloneTunerUtil.readTunerValByQuery("deadband and base and cooling and " +
-                "equipRef == \"" + q.getId() + "\"");
-        double hdb = StandaloneTunerUtil.readTunerValByQuery("deadband and base and heating and " +
-                "equipRef == \"" + q.getId() + "\"");
         String zoneId = HSUtil.getZoneIdFromEquipId(q.getId());
         Occupied occ = ScheduleProcessJob.getOccupiedModeCache(zoneId);
-        if (occ != null) {
-            cdb = occ.getCoolingDeadBand();
-            hdb = occ.getHeatingDeadBand();
-        }
-        double coolingDesiredTemp = dt + cdb;
-        double heatingDesiredTemp = dt - hdb;
-
-
-        HashMap coolingDtPoint = CCUHsApi.getInstance().read("point and air and temp and " +
+        HashMap<Object,Object> coolingDtPoint = hayStack.readEntity("point and air and temp and " +
                 "desired and cooling and sp and equipRef == \"" + q.getId() + "\"");
         if (coolingDtPoint == null || coolingDtPoint.size() == 0) {
             throw new IllegalArgumentException();
         }
-        try {
-            CCUHsApi.getInstance().writeHisValById(coolingDtPoint.get("id").toString(),
-                    coolingDesiredTemp);
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
-
-        HashMap heatinDtPoint = CCUHsApi.getInstance().read("point and air and temp and " +
+        HashMap<Object,Object> heatinDtPoint = hayStack.readEntity("point and air and temp and " +
                 "desired and heating and sp and equipRef == \"" + q.getId() + "\"");
         if (heatinDtPoint == null || heatinDtPoint.size() == 0) {
             throw new IllegalArgumentException();
         }
-
-        try {
-            CCUHsApi.getInstance().writeHisValById(heatinDtPoint.get("id").toString(),
-                    heatingDesiredTemp);
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
-
-        HashMap singleDtPoint = CCUHsApi.getInstance().read("point and air and temp and " +
+        HashMap<Object,Object>  avgDtPoint = CCUHsApi.getInstance().readEntity("point and air and temp and " +
                 "desired and average and sp and equipRef == \"" + q.getId() + "\"");
-        if (singleDtPoint == null || singleDtPoint.size() == 0) {
+        if (avgDtPoint == null || avgDtPoint.size() == 0) {
             throw new IllegalArgumentException();
         }
-        try {
-            CCUHsApi.getInstance().writeHisValById(singleDtPoint.get("id").toString(), dt);
-        } catch (Exception e) {
-            e.printStackTrace();
+        double coolingDesiredTemp;
+        double heatingDesiredTemp;
+        if (occ != null) {
+            if(isFirstTime){
+                coolingDesiredTemp = occ.getCoolingVal();
+                heatingDesiredTemp = occ.getHeatingVal();
+            }else{
+                coolingDesiredTemp = hayStack.readPointPriorityVal(Objects.requireNonNull(coolingDtPoint.get("id")).toString());
+                heatingDesiredTemp = hayStack.readPointPriorityVal(Objects.requireNonNull(heatinDtPoint.get("id")).toString());
+            }
+
+            CCUHsApi.getInstance().writeHisValById(Objects.requireNonNull(coolingDtPoint.get("id")).toString(),
+                    coolingDesiredTemp);
+            CCUHsApi.getInstance().writeHisValById(Objects.requireNonNull(heatinDtPoint.get("id")).toString(),
+                    heatingDesiredTemp);
+            CCUHsApi.getInstance().writeHisValById(Objects.requireNonNull(avgDtPoint.get("id")).toString(),
+                    avgDesiredTemp);
+            SystemScheduleUtil.handleManualDesiredTempUpdate(new Point.Builder().setHashMap(coolingDtPoint).build(),
+                    new Point.Builder().setHashMap(heatinDtPoint).build(),
+                    new Point.Builder().setHashMap(avgDtPoint).build(),
+                    coolingDesiredTemp,
+                    heatingDesiredTemp,
+                    avgDesiredTemp);
         }
-        SystemScheduleUtil.handleManualDesiredTempUpdate(new Point.Builder().setHashMap(coolingDtPoint).build(),
-                                                         new Point.Builder().setHashMap(heatinDtPoint).build(),
-                                                         new Point.Builder().setHashMap(singleDtPoint).build(),
-                                                         coolingDesiredTemp,
-                                                         heatingDesiredTemp,
-                                                         dt);
+
+
 
     }
 
     private void runAutoAwayOperation() {
-        HashMap equipMap =
-                CCUHsApi.getInstance().read("equip and group == \"" + mBPOSEquip.mNodeAddr + "\"");
+        HashMap<Object,Object> equipMap =
+                CCUHsApi.getInstance().readEntity("equip and group == \"" + mBPOSEquip.mNodeAddr + "\"");
         Equip q = new Equip.Builder().setHashMap(equipMap).build();
         String zoneId = HSUtil.getZoneIdFromEquipId(mBPOSEquip.mEquipRef);
         Occupied occ = ScheduleProcessJob.getOccupiedModeCache(zoneId);
-        boolean occupied = (occ == null ? false : occ.isOccupied());
-        HashMap occupancymode = CCUHsApi.getInstance().read(
+        boolean occupied = (occ != null && occ.isOccupied());
+        HashMap<Object,Object> occupancymode = CCUHsApi.getInstance().readEntity(
                 "point and occupancy  and mode and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
         double occupancyModeval = CCUHsApi.getInstance().readHisValByQuery(
                 "point and  bpos and occupancy  and his and " +
@@ -408,83 +369,83 @@ public class BPOSProfile extends ZoneProfile {
         boolean occupancysensor = CCUHsApi.getInstance().readHisValByQuery(
                 "point and  bpos and occupancy  and his and " +
                         "sensor and equipRef  == \"" + mBPOSEquip.mEquipRef + "\"") > 0;
-        HashMap ocupancyDetection = CCUHsApi.getInstance().read(
+        HashMap<Object,Object> ocupancyDetection = CCUHsApi.getInstance().readEntity(
                 "point and  bpos and occupancy and detection and his and equipRef  == \"" + mBPOSEquip.mEquipRef + "\"");
 
         if (occupied && occupancyModeval != Occupancy.AUTOFORCEOCCUPIED.ordinal()) {
-            Log.d("BPOSProfile", "  auto away handle");
+            Log.d(L.TAG_BPOS ,"  auto away handle");
             //if the oocupantnotdetected for autoAwayTimer
             // then enter autoaway
             if (!occupancysensor) {
-                Log.d("BPOSProfile", "  auto away case -  occDetPoint <= 0");
+                Log.d(L.TAG_BPOS, "  auto away case -  occDetPoint <= 0");
                 //check if there was no detection from past autoawaytimer
                 HisItem hisItem =
-                        CCUHsApi.getInstance().curRead(ocupancyDetection.get("id").toString());
+                        CCUHsApi.getInstance().curRead(Objects.requireNonNull(ocupancyDetection.get("id")).toString());
                 long lastupdatedtime = (hisItem == null) ? null : hisItem.getDate().getTime();
-                DateTime et = new DateTime(lastupdatedtime);
-
-
                 long min = findDifference(lastupdatedtime, false);
                 // read autoawaytimer tuner
                 double autoawaytime = TunerUtil.readTunerValByQuery(" point and tuner and auto " +
                         "and away and " +
                         "time ", mBPOSEquip.mEquipRef);
 
-                Log.d("BPOSProfile", "  auto away case -  min = " + min + " autoawaytime ="
+                Log.d(L.TAG_BPOS, "  auto away case -  min = " + min + " autoawaytime ="
                         + autoawaytime + "lastupdatedtime" + lastupdatedtime);
 
                 if (min >= autoawaytime && occupancyModeval != (double) Occupancy.AUTOAWAY.ordinal()) {
-                    HashMap coolingDtPoint = CCUHsApi.getInstance().read("point and air and temp and " +
+                    HashMap<Object,Object> coolingDtPoint = CCUHsApi.getInstance().readEntity("point and air and temp and " +
                             "desired and cooling and sp and equipRef == \"" + q.getId() + "\"");
                     if (coolingDtPoint == null || coolingDtPoint.size() == 0) {
                         throw new IllegalArgumentException();
                     }
-                    HashMap heatinDtPoint = CCUHsApi.getInstance().read("point and air and temp and " +
+                    HashMap<Object,Object> heatinDtPoint = CCUHsApi.getInstance().readEntity("point and air and temp and " +
                             "desired and heating and sp and equipRef == \"" + q.getId() + "\"");
                     if (heatinDtPoint == null || heatinDtPoint.size() == 0) {
                         throw new IllegalArgumentException();
                     }
-                    //SystemScheduleUtil.clearOverrides(coolingDtPoint.get("id").toString());
-                    //SystemScheduleUtil.clearOverrides(heatinDtPoint.get("id").toString());
-                    CCUHsApi.getInstance().writeHisValById(occupancymode.get("id").toString(),
+                    CCUHsApi.getInstance().writeHisValById(Objects.requireNonNull(occupancymode.get("id")).toString(),
                             (double) Occupancy.AUTOAWAY.ordinal());
-                    CCUHsApi.getInstance().writeHisValById(ocupancyDetection.get("id").toString(),
+                    CCUHsApi.getInstance().writeHisValById(Objects.requireNonNull(ocupancyDetection.get("id")).toString(),
                             0.0);
                 }
             } else {
-                if (occupancysensor)
-                    CCUHsApi.getInstance().writeHisValById(occupancymode.get("id").toString(),
+                if (occupancyModeval != Occupancy.OCCUPIED.ordinal()) {
+                    CCUHsApi.getInstance().writeHisValById(Objects.requireNonNull(occupancymode.get(
+                            "id")).toString(),
                             (double) Occupancy.OCCUPIED.ordinal());
-                clearlevel3Override();
-            }
-        }else{
-            Log.d("BPOSProfile", "in else");
-            if (ScheduleProcessJob.getSystemOccupancy() == Occupancy.PRECONDITIONING) {
-                CCUHsApi.getInstance().writeHisValById(occupancymode.get("id").toString(),
-                        (double) Occupancy.PRECONDITIONING.ordinal());
-            } else {
-                CCUHsApi.getInstance().writeHisValById(occupancymode.get("id").toString(),
-                        (double) Occupancy.UNOCCUPIED.ordinal());
+                    clearlevel3Override();
+                }
             }
+        }else {
+            Log.d(L.TAG_BPOS, "  auto away case -  reset to occupied");
             clearlevel3Override();
+
         }
     }
 
     private void clearlevel3Override() {
-        HashMap coolDT = CCUHsApi.getInstance().read("point and desired and cooling and temp and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
-        HashMap heatDT = CCUHsApi.getInstance().read("point and desired and heating and temp and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
-        HashMap averageDT = CCUHsApi.getInstance().read("point and desired and average and temp and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
-        double coolDTl3 =  HSUtil.getPriorityLevelVal(coolDT.get("id").toString(),3);
-        double heatDTl3 =  HSUtil.getPriorityLevelVal(coolDT.get("id").toString(),3);
-
+        HashMap<Object,Object> coolDT = CCUHsApi.getInstance().readEntity("point and desired and cooling and temp and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
+        HashMap<Object,Object> heatDT = CCUHsApi.getInstance().readEntity("point and desired and heating and temp and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
+        HashMap<Object,Object> averageDT = CCUHsApi.getInstance().readEntity("point and desired and average and temp and equipRef == \"" + mBPOSEquip.mEquipRef + "\"");
+        double coolDTl3 =  HSUtil.getPriorityLevelVal(Objects.requireNonNull(coolDT.get("id")).toString(),3);
+        double heatDTl3 =  HSUtil.getPriorityLevelVal(Objects.requireNonNull(heatDT.get("id")).toString(),3);
+        double avgDT =  HSUtil.getPriorityLevelVal(Objects.requireNonNull(averageDT.get("id")).toString(),3);
 
         if( coolDTl3 != 0.0) {
-            CCUHsApi.getInstance().pointWrite(HRef.copy(coolDT.get("id").toString()), 3, "manual", HNum.make(0), HNum.make(1, "ms"));
+            CCUHsApi.getInstance().pointWrite(HRef.copy(Objects.requireNonNull(coolDT.get("id")).toString()), 3, "manual", HNum.make(0), HNum.make(1, "ms"));
         }
         if( heatDTl3 != 0.0) {
-            CCUHsApi.getInstance().pointWrite(HRef.copy(heatDT.get("id").toString()), 3, "manual"
+            CCUHsApi.getInstance().pointWrite(HRef.copy(Objects.requireNonNull(heatDT.get("id")).toString()), 3, "manual"
                     , HNum.make(0), HNum.make(1, "ms"));
         }
+        if (avgDT != 0.0) {
+            CCUHsApi.getInstance().pointWrite(
+                    HRef.copy(Objects.requireNonNull(averageDT.get("id")).toString()),
+                    3,
+                    "manual",
+                    HNum.make(0),
+                    HNum.make(1, "ms")
+            );
+        }
 
 
     }
@@ -492,7 +453,7 @@ public class BPOSProfile extends ZoneProfile {
     @Override
     public Equip getEquip()
     {
-        HashMap equip = CCUHsApi.getInstance().read("equip and group == \""+mBPOSEquip.mNodeAddr+"\"");
+        HashMap<Object,Object> equip = CCUHsApi.getInstance().readEntity("equip and group == \""+mBPOSEquip.mNodeAddr+"\"");
         return new Equip.Builder().setHashMap(equip).build();
     }
 
@@ -502,16 +463,18 @@ public class BPOSProfile extends ZoneProfile {
                 "auto and forced and away and config and equipRef == \"" + mBPOSEquip.mEquipRef + "\"") > 0;
         Occupied occuStatus =
                 ScheduleProcessJob.getOccupiedModeCache(HSUtil.getZoneIdFromEquipId(mBPOSEquip.mEquipRef ));
-        if(isAutoawayenabled && occuStatus.isOccupied()){
-            HashMap ocupancyDetection = CCUHsApi.getInstance().read(
-                    "point and  bpos and occupancy and detection and his and equipRef  ==" +
-                            " \"" + mBPOSEquip.mEquipRef  + "\"");
-            if (ocupancyDetection.get("id") != null) {
-                double val = CCUHsApi.getInstance().readHisValById(ocupancyDetection.get(
-                        "id").toString());
-                CCUHsApi.getInstance().writeHisValueByIdWithoutCOV(ocupancyDetection.get(
-                        "id").toString(),
-                        val);
+        if(isAutoawayenabled && occuStatus != null) {
+            if (occuStatus.isOccupied()) {
+                HashMap<Object,Object> ocupancyDetection = CCUHsApi.getInstance().readEntity(
+                        "point and  bpos and occupancy and detection and his and equipRef  ==" +
+                                " \"" + mBPOSEquip.mEquipRef + "\"");
+                if (ocupancyDetection.get("id") != null) {
+                    double val = CCUHsApi.getInstance().readHisValById(Objects.requireNonNull(ocupancyDetection.get(
+                            "id")).toString());
+                    CCUHsApi.getInstance().writeHisValueByIdWithoutCOV(Objects.requireNonNull(ocupancyDetection.get(
+                            "id")).toString(),
+                            val);
+                }
             }
         }
     }
diff --git a/logic/src/main/java/a75f/io/logic/jobs/ScheduleProcessJob.java b/logic/src/main/java/a75f/io/logic/jobs/ScheduleProcessJob.java
index a2b35616d..de3c7db58 100644
--- a/logic/src/main/java/a75f/io/logic/jobs/ScheduleProcessJob.java
+++ b/logic/src/main/java/a75f/io/logic/jobs/ScheduleProcessJob.java
@@ -1538,7 +1538,7 @@ public class ScheduleProcessJob extends BaseJob implements WatchdogMonitor
                         clearTempOverrides(equip.getId());
                     } else if (prevStatus == AUTOFORCEOCCUPIED){
                         occupancyState = AUTOFORCEOCCUPIED;
-                    }else {
+                    } else if(prevStatus == FORCEDOCCUPIED) {
                         occupancyState = FORCEDOCCUPIED;
                     }
                 } else if ((cachedOccupied != null) && cachedOccupied.getVacation() != null) {
-- 
2.31.0.windows.1

